File: src/main/java/me/nasukhov/bot/command/AnswerQuestion.java
package me.nasukhov.bot.command;

import me.nasukhov.bot.*;
import me.nasukhov.study.ChannelQuestion;
import me.nasukhov.study.ProgressRepository;

import java.util.*;

public class AnswerQuestion implements Handler {
    private static final String ANSWER_CORRECT = "%s, правильно «%s».\n• ᴗ •";
    private static final String ANSWER_INCORRECT = "%s, правильно «%s».\n• ᴖ •";
    private final ProgressRepository progressRepository;
    private final AskQuestion askQuestion;

    public AnswerQuestion(ProgressRepository progressRepository, AskQuestion askQuestion) {
        this.progressRepository = progressRepository;
        this.askQuestion = askQuestion;
    }

    @Override
    public boolean supports(Input input) {
        return input.input().startsWith("qh answer ");
    }

    @Override
    public void handle(Input input, Output output) {
        String[] parts = input.input().split("\\s+", 4);
        if (parts.length < 4) {
            return;
        }

        UUID channelQuestionId;
        try {
            channelQuestionId = UUID.fromString(parts[2]);
        } catch (IllegalArgumentException e) {
            return;
        }

        Channel channel = input.channel();
        String userId = input.sender().id();

        boolean alreadyAnswered = progressRepository.hasReplyInChannel(userId, channel.id, channelQuestionId);
        if (alreadyAnswered) {
            return;
        }

        Optional<ChannelQuestion> result = progressRepository.findQuestionInChannel(channelQuestionId);;
        if (result.isEmpty()) {
            return;
        }

        ChannelQuestion question = result.get();

        int selectedOption = Integer.parseInt(parts[3]);
        boolean isCorrectAnswer = question.isCorrectAnswer(selectedOption);
        progressRepository.addUserAnswer(
                channelQuestionId,
                userId,
                channel.id,
                isCorrectAnswer
        );

        String template = isCorrectAnswer ? ANSWER_CORRECT: ANSWER_INCORRECT;
        output.write(String.format(template, input.sender().name(), question.viewAnswer()));

        if (!channel.isPublic()) {
            this.askQuestion.handle(input, output);
        }
    }
}

File: src/main/kotlin/me/nasukhov/tukitalearner/bot/AnswerQuestion.kt
package me.nasukhov.tukitalearner.bot

import me.nasukhov.bot.command.Handler
import me.nasukhov.bot.io.Input
import me.nasukhov.bot.io.NoOutput
import me.nasukhov.bot.io.Output
import me.nasukhov.bot.io.TextOutput
import me.nasukhov.tukitalearner.study.GroupQuestionRepository
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

private data class AnswerData(
    val questionId: UUID,
    val selectedOption: Int,
)

@Component
class AnswerQuestion(
    private val groupQuestionRepository: GroupQuestionRepository,
) : Handler {
    companion object {
        private const val ANSWER_CORRECT = "%s, да, правильный ответ «<spoiler>%s</spoiler>»."
        private const val ANSWER_INCORRECT = "%s, увы, правильный ответ «<spoiler>%s</spoiler>»."
    }

    override fun supports(input: Input): Boolean = input.raw.startsWith("qh answer ")

    @Transactional
    override fun handle(input: Input): Output {
        if (!supports(input)) {
            return NoOutput()
        }

        val answerData = parseGroupQuestionId(input.raw)
        if (answerData == null) {
            return NoOutput()
        }

        val result = groupQuestionRepository.findByIdWithAnswers(answerData.questionId)
        if (result.isEmpty) {
            return NoOutput()
        }

        val userId = input.sender.id
        val question = result.get()
        if (question.hasAnswerFromUser(userId)) {
            return NoOutput()
        }

        val answer = question.addAnswer(userId, answerData.selectedOption)

        val template: String = if (answer.isCorrect) ANSWER_CORRECT else ANSWER_INCORRECT

        return TextOutput(
            String.format(
                template,
                input.sender.name,
                question.viewAnswer(),
            ),
        )
    }

    private fun parseGroupQuestionId(input: String): AnswerData? {
        val parts = input.split("\\s+".toRegex(), limit = 4).toTypedArray()
        if (parts.size < 4) {
            return null
        }

        return try {
            AnswerData(UUID.fromString(parts[2]), parts[3].toInt())
        } catch (_: IllegalArgumentException) {
            null
        }
    }
}
